File: apps/app/lib/contracts/v2/lifecycle.ts
/**
 * F2C.a — the console half of `infrx/contracts/v2/lifecycle.py`.
 *
 * Browser-safe only: the vocabularies, the refusal -> code map, the upload ticket (it
 * carries no object key) and the readiness VIEW. The content rows, claims, tombstones and
 * manifests carry server object keys and stay in Python; no browser receives them.
 *
 * The decoders are exact: a missing required field, an unexpected field, a wrong schema
 * version or a broken invariant throws a `TypeError`, deterministically, so an old or
 * foreign payload can never be read as an "empty" or "available" one. Fixtures are read
 * from the Python package's `fixtures/v2/` (one source, no copy).
 *
 * Owned by the coordinator: a change here is a contract revision.
 */

import { instantKey } from "./types.ts";

export const UPLOAD_STATES = ["created", "finalized", "aborted", "expired"] as const;
export type UploadState = (typeof UPLOAD_STATES)[number];

/** `not_ready` = no committed marker (never completed). NOT the empty manifest. */
export const READINESS_STATES = ["not_ready", "ready"] as const;
export type ReadinessState = (typeof READINESS_STATES)[number];

export const CONTENT_KINDS = [
  "upload_destination",
  "source",
  "prepared",
  "payload",
  "result",
] as const;
export type ContentKind = (typeof CONTENT_KINDS)[number];

export const CONTENT_LOCATIONS = ["object_store", "database"] as const;
export const CONTENT_ORIGINS = ["written", "discovered"] as const;
export const LIFECYCLE_STATES = ["live", "tombstoned", "deleted"] as const;
export type LifecycleState = (typeof LIFECYCLE_STATES)[number];

export const LIFECYCLE_REFUSALS = [
  "not_found",
  "invalid_constraints",
  "upload_expired",
  "upload_not_open",
  "upload_not_finalized",
  "nothing_received",
  "bytes_changed",
  "too_large",
  "size_mismatch",
  "digest_mismatch",
  "mime_not_accepted",
  "invalid_manifest",
  "expectation_mismatch",
  "not_ready",
  "content_retiring",
  "not_eligible",
  "reference_live",
  "claim_held",
  "claim_lost",
] as const;
export type LifecycleRefusal = (typeof LIFECYCLE_REFUSALS)[number];

/** Reason -> the existing error code it is raised as. No new public code exists. */
export const LIFECYCLE_REFUSAL_CODES: Readonly<Record<LifecycleRefusal, string>> = Object.freeze({
  not_found: "not_found",
  invalid_constraints: "invalid_request",
  upload_expired: "upload_expired",
  upload_not_open: "state_conflict",
  upload_not_finalized: "invalid_request",
  nothing_received: "invalid_request",
  bytes_changed: "state_conflict",
  too_large: "request_too_large",
  size_mismatch: "invalid_request",
  digest_mismatch: "unsupported_media",
  mime_not_accepted: "unsupported_media",
  invalid_manifest: "invalid_request",
  expectation_mismatch: "invalid_request",
  not_ready: "not_claimable",
  content_retiring: "dependency_unavailable",
  not_eligible: "not_claimable",
  reference_live: "not_claimable",
  claim_held: "not_claimable",
  claim_lost: "stale_lease",
});

export const DESTINATION_SCHEME = "infrx-upload:";
const V2 = 2;
const UUID = /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/;
const UPLOAD_HANDLE = /^upl_[A-Za-z0-9_-]{22,64}$/;
const SHA256 = /^sha256:[0-9a-f]{64}$/;

export type UploadConstraints = {
  schema_version: 2;
  max_bytes: number;
  bytes?: number;
  accepted_mime: readonly string[];
  digest?: string;
};
export type UploadReceipt = { schema_version: 2; bytes: number; digest: string; received_at: string };
export type FinalizedSource = {
  schema_version: 2;
  content_id: string;
  generation: number;
  digest: string;
  bytes: number;
  mime: string;
  profile_version: string;
  duration_s: number;
  finalized_at: string;
};
export type UploadTicket = {
  schema_version: 2;
  upload_handle: string;
  org_id: string;
  destination_ref: string;
  constraints: UploadConstraints;
  state: UploadState;
  created_at: string;
  expires_at: string;
  received?: UploadReceipt;
  finalized?: FinalizedSource;
  refusal?: LifecycleRefusal;
};
export type ReadinessView = {
  schema_version: 2;
  request_id: string;
  state: ReadinessState;
  source_count?: number;
  ready_at?: string;
};

// --- exact decoding ------------------------------------------------------------------------
type Obj = Record<string, unknown>;

function fail(what: string): never {
  throw new TypeError(`contract violation: ${what}`);
}

/** The object with exactly `required` (all present) plus any of `optional`, and v2. */
function exact(value: unknown, what: string, required: string[], optional: string[] = []): Obj {
  if (typeof value !== "object" || value === null || Array.isArray(value)) fail(`${what} is not an object`);
  const obj = value as Obj;
  for (const key of ["schema_version", ...required]) {
    if (!(key in obj) || obj[key] === undefined || obj[key] === null) fail(`${what}.${key} is missing`);
  }
  const allowed = new Set(["schema_version", ...required, ...optional]);
  for (const key of Object.keys(obj)) if (!allowed.has(key)) fail(`${what}.${key} is unexpected`);
  if (obj.schema_version !== V2) fail(`${what}.schema_version is not 2`);
  for (const key of optional) if (key in obj && obj[key] === null) fail(`${what}.${key} is null`);
  return obj;
}

function text(obj: Obj, key: string, pattern?: RegExp): string {
  const value = obj[key];
  if (typeof value !== "string" || value.length === 0 || (pattern && !pattern.test(value))) {
    fail(`${key} is not a valid string`);
  }
  return value;
}

function count(obj: Obj, key: string, min: number): number {
  const value = obj[key];
  if (typeof value !== "number" || !Number.isInteger(value) || value < min) fail(`${key} is not an integer >= ${min}`);
  return value;
}

function instant(obj: Obj, key: string): string {
  const value = text(obj, key);
  instantKey(value); // throws on anything but a UTC contract instant
  return value;
}

function member<T extends string>(obj: Obj, key: string, allowed: readonly T[]): T {
  const value = obj[key];
  if (typeof value !== "string" || !(allowed as readonly string[]).includes(value)) fail(`${key} is not one of ${allowed.join("|")}`);
  return value as T;
}

function before(a: string, b: string): boolean {
  return instantKey(a) < instantKey(b);
}

function decodeConstraints(value: unknown): UploadConstraints {
  const obj = exact(value, "constraints", ["max_bytes", "accepted_mime"], ["bytes", "digest"]);
  const max = count(obj, "max_bytes", 1);
  if ("bytes" in obj && count(obj, "bytes", 1) > max) fail("declared bytes exceed max_bytes");
  if ("digest" in obj) text(obj, "digest", SHA256);
  const mimes = obj.accepted_mime;
  if (!Array.isArray(mimes) || mimes.length === 0 || new Set(mimes).size !== mimes.length
      || !mimes.every((m) => typeof m === "string" && m.length > 0 && m === m.trim().toLowerCase())) {
    fail("accepted_mime is a list of distinct lowercase media types");
  }
  return obj as UploadConstraints;
}

/** An upload ticket, with every rule `UploadTicket._one_fact` enforces in Python. */
export function decodeUploadTicket(value: unknown): UploadTicket {
  const obj = exact(value, "ticket",
    ["upload_handle", "org_id", "destination_ref", "constraints", "state", "created_at", "expires_at"],
    ["received", "finalized", "refusal"]);
  const handle = text(obj, "upload_handle", UPLOAD_HANDLE);
  text(obj, "org_id", UUID);
  if (obj.destination_ref !== DESTINATION_SCHEME + handle) fail("destination_ref is infrx-upload:<handle>");
  const c = decodeConstraints(obj.constraints);
  const state = member(obj, "state", UPLOAD_STATES);
  const created = instant(obj, "created_at");
  const expires = instant(obj, "expires_at");
  if (!before(created, expires)) fail("expires_at must follow created_at");
  if ((state === "aborted") !== ("refusal" in obj)) fail("aborted exactly when a refusal is recorded");
  if ("refusal" in obj) member(obj, "refusal", LIFECYCLE_REFUSALS);
  let received: Obj | undefined;
  if ("received" in obj) {
    received = exact(obj.received, "received", ["bytes", "digest", "received_at"]);
    if (count(received, "bytes", 0) > c.max_bytes) fail("received over max_bytes");
    text(received, "digest", SHA256);
    const at = instant(received, "received_at");
    if (before(at, created) || !before(at, expires)) fail("received outside the window");
  }
  if ((state === "finalized") !== ("finalized" in obj)) fail("finalized exactly when a source is recorded");
  if ("finalized" in obj) {
    const done = exact(obj.finalized, "finalized", ["content_id", "generation", "digest", "bytes", "mime",
      "profile_version", "duration_s", "finalized_at"]);
    text(done, "content_id", UUID);
    count(done, "generation", 1);
    const digest = text(done, "digest", SHA256);
    const bytes = count(done, "bytes", 0);
    const mime = text(done, "mime");
    text(done, "profile_version");
    if (typeof done.duration_s !== "number" || !(done.duration_s >= 0)) fail("duration_s is a measured number");
    const at = instant(done, "finalized_at");
    if (!received || received.digest !== digest || received.bytes !== bytes) fail("a finalized source is exactly the bytes received");
    if (!c.accepted_mime.includes(mime) || (c.bytes !== undefined && c.bytes !== bytes)
        || (c.digest !== undefined && c.digest !== digest)) fail("a finalized source satisfies every constraint");
    if (before(at, received.received_at as string) || !before(at, expires)) fail("finalized outside the window");
  }
  return obj as UploadTicket;
}

/** `ready` carries a count (possibly 0) and an instant; `not_ready` neither. */
export function decodeReadinessView(value: unknown): ReadinessView {
  const obj = exact(value, "readiness", ["request_id", "state"], ["source_count", "ready_at"]);
  text(obj, "request_id", UUID);
  const ready = member(obj, "state", READINESS_STATES) === "ready";
  if (ready !== ("source_count" in obj) || ready !== ("ready_at" in obj)) {
    fail("ready carries source_count and ready_at; not_ready neither");
  }
  if (ready) {
    count(obj, "source_count", 0);
    instant(obj, "ready_at");
  }
  return obj as ReadinessView;
}

/** A finalized upload is usable while `now < expires_at`; at equality it has expired. */
export function uploadUsable(ticket: UploadTicket, now: string): boolean {
  return ticket.state === "finalized" && before(now, ticket.expires_at);
}

File: apps/infrx-api/infrx/contracts/fakes/lifecycle.py
"""The reference adapter of `contracts.v2.lifecycle`'s three ports, in memory.

`FakeLifecycle` is one store the way D10's PostgreSQL is one database: tickets, content
rows, references and readiness markers live in a shared `Durable` record, and
`reopen()` is another process over the same state (nothing is kept per instance, which is
exactly what `MediaUploads.uploads` and `MediaCollector` got wrong - RV-02/RV-03). One
`asyncio.Lock` over `Durable` stands for the row locks that serialize admission,
completion and deletion.

Admission itself is the existing `FakeJobStore` (`jobs`): `admit_ready` runs every
lifecycle check first, then `admit`/`admit_credit`, then records the marker, all inside
the lock with no await that yields to another lifecycle operation, so it is one
transaction as far as any case can observe.
"""
from __future__ import annotations

import asyncio
import base64
import json
import re
from dataclasses import dataclass, field
from datetime import datetime, timedelta

from .. import errors, ids
from ..limits import DEFAULTS
from ..records import ExecutionMode, MediaKind, MediaRef, UploadState
from ..v2.lifecycle import (MAX_PAGE, AdmissionExpectation, ContentIdentity, ContentKind,
                            ContentLocation, ContentObject, ContentOrigin, ContentPage,
                            ContentReference, DeletionClaim, ExecutionReadiness,
                            FinalizedSource, LifecycleRefusal as R, LifecycleState,
                            ManifestSource, Tombstone, UploadConstraints, UploadReceipt,
                            UploadTicket, DESTINATION_SCHEME, UPLOAD_ABORT_REASONS, refuse)
from ..v2.records import AccountingRegime, AdmissionV2

UPLOAD_TTL_S = DEFAULTS.processing_cache_ttl_s        # the window MediaUploads uses today
GRACE_S = DEFAULTS.processing_cache_ttl_s             # MediaCollector's grace today
CLAIM_TTL_S = 60.0
RETENTION_S = DEFAULTS.processing_cache_ttl_s         # P-25 pending: a fixture value only
SHA256 = re.compile(r"sha256:[0-9a-f]{64}")


@dataclass
class Durable:
    """Everything a restart must not lose."""

    tickets: dict[str, UploadTicket] = field(default_factory=dict)
    content: dict[str, ContentObject] = field(default_factory=dict)
    by_key: dict[tuple[str, str], str] = field(default_factory=dict)
    references: dict[str, list[ContentReference]] = field(default_factory=dict)
    readiness: dict[str, ExecutionReadiness] = field(default_factory=dict)
    retain: dict[tuple, datetime] = field(default_factory=dict)   # (job, kind) -> until
    lock: asyncio.Lock = field(default_factory=asyncio.Lock)


class FakeLifecycle:
    def __init__(self, jobs, clock, ids_, *, durable: Durable | None = None,
                 upload_ttl_s: float = UPLOAD_TTL_S, grace_s: float = GRACE_S,
                 claim_ttl_s: float = CLAIM_TTL_S, retention_s: float = RETENTION_S) -> None:
        self.jobs, self.clock, self.ids = jobs, clock, ids_
        self.d = durable or Durable()
        self.upload_ttl_s, self.grace_s = upload_ttl_s, grace_s
        self.claim_ttl_s, self.retention_s = claim_ttl_s, retention_s

    def reopen(self) -> FakeLifecycle:
        """Another process over the same durable state."""
        return FakeLifecycle(self.jobs, self.clock, self.ids, durable=self.d,
                             upload_ttl_s=self.upload_ttl_s, grace_s=self.grace_s,
                             claim_ttl_s=self.claim_ttl_s, retention_s=self.retention_s)

    def _put(self, ticket: UploadTicket, **changes) -> UploadTicket:
        # Revalidated, never `model_copy(update=)` (R78): the record's own rules hold.
        ticket = UploadTicket.model_validate({**ticket.model_dump(), **changes})
        self.d.tickets[ticket.upload_handle] = ticket
        return ticket

    def _owned(self, org_id: str, handle: object) -> UploadTicket:
        ticket = self.d.tickets.get(handle) if isinstance(handle, str) else None
        if ticket is None or ticket.org_id != org_id:
            raise refuse(R.not_found, f"no upload {handle!r} for this organization")
        return ticket

    def _open(self, ticket: UploadTicket) -> None:
        if ticket.state is not UploadState.created:
            raise refuse(R.upload_not_open, f"upload is {ticket.state}")
        if self.clock.now() >= ticket.expires_at:
            raise refuse(R.upload_expired, "the upload window closed")

    # --- UploadRepository ----------------------------------------------------------------
    async def create(self, org_id: str, constraints: UploadConstraints) -> UploadTicket:
        if not ids.is_request_id(org_id):
            raise errors.InvalidRequest("org_id must be a lowercase UUID")
        if not isinstance(constraints, UploadConstraints):
            raise refuse(R.invalid_constraints, "constraints are an UploadConstraints record")
        async with self.d.lock:
            now, handle = self.clock.now(), self.ids.upload_handle()
            return self._put(UploadTicket(
                upload_handle=handle, org_id=org_id, destination_ref=DESTINATION_SCHEME + handle,
                constraints=constraints, state=UploadState.created, created_at=now,
                expires_at=now + timedelta(seconds=self.upload_ttl_s)))

    async def acknowledge_put(self, org_id: str, upload_handle: str, *, bytes: int,
                              digest: str) -> UploadTicket:
        async with self.d.lock:
            ticket = self._owned(org_id, upload_handle)
            self._open(ticket)
            if isinstance(bytes, bool) or not isinstance(bytes, int) or bytes < 0:
                raise errors.InvalidRequest("bytes is a nonnegative integer")
            if bytes > ticket.constraints.max_bytes:
                raise refuse(R.too_large, "over the ticket's max_bytes")
            if not isinstance(digest, str) or not SHA256.fullmatch(digest):
                raise errors.InvalidRequest("digest is sha256:<64 hex>")
            receipt = UploadReceipt(bytes=bytes, digest=digest, received_at=self.clock.now())
            if ticket.received is not None:
                if (ticket.received.bytes, ticket.received.digest) != (bytes, digest):
                    raise refuse(R.bytes_changed, "the handle already received other bytes")
                return ticket
            return self._put(ticket, received=receipt)

    async def complete(self, org_id: str, upload_handle: str, source: MediaRef) -> UploadTicket:
        async with self.d.lock:
            ticket = self._owned(org_id, upload_handle)
            if (source.org_id, source.handle, source.kind) != (org_id, upload_handle,
                                                               MediaKind.upload):
                raise refuse(R.not_found, "the source is not this ticket's")
            done = ticket.finalized
            if done is not None:
                row = self.d.content.get(done.content_id)
                if (done.digest, done.bytes, done.mime) != (source.digest, source.bytes,
                                                           source.mime) \
                        or row is None or row.identity.object_key != source.storage_ref:
                    raise refuse(R.bytes_changed, "finalized with other bytes")
                return ticket
            self._open(ticket)
            received, c = ticket.received, ticket.constraints
            if source.duration_s is None:
                raise errors.InvalidRequest("a finalized upload carries its measured duration")
            if received is None:
                raise refuse(R.nothing_received, "no bytes reached the destination")
            if (source.digest, source.bytes) != (received.digest, received.bytes):
                raise refuse(R.bytes_changed, "the destination changed since it was received")
            failed = (R.size_mismatch if c.bytes is not None and source.bytes != c.bytes
                      else R.digest_mismatch if c.digest is not None and source.digest != c.digest
                      else R.mime_not_accepted if source.mime not in c.accepted_mime else None)
            if failed is not None:
                self._put(ticket, state=UploadState.aborted, refusal=failed)
                raise refuse(failed, "a constraint failed; the ticket is aborted")
            row = self._register(ContentIdentity(
                org_id=org_id, kind=ContentKind.source, location=ContentLocation.object_store,
                object_key=source.storage_ref, digest=source.digest, bytes=source.bytes,
                origin=ContentOrigin.written))
            return self._put(ticket, state=UploadState.finalized, finalized=FinalizedSource(
                content_id=row.content_id, generation=row.generation, digest=source.digest,
                bytes=source.bytes, mime=source.mime, profile_version=source.profile_version,
                duration_s=source.duration_s, finalized_at=self.clock.now()))

    async def abort(self, org_id: str, upload_handle: str, refusal: R) -> UploadTicket:
        if refusal not in UPLOAD_ABORT_REASONS:
            raise errors.InvalidRequest("a ticket records only a public upload refusal")
        async with self.d.lock:
            ticket = self._owned(org_id, upload_handle)
            if ticket.state is UploadState.aborted:
                return ticket
            if ticket.state is not UploadState.created or self.clock.now() >= ticket.expires_at:
                raise refuse(R.upload_not_open, f"upload is {ticket.state} or past its window")
            return self._put(ticket, state=UploadState.aborted, refusal=R(refusal))

    async def resolve(self, org_id: str, upload_handle: str) -> UploadTicket:
        async with self.d.lock:
            return self._resolve(org_id, upload_handle)

    def _resolve(self, org_id: str, upload_handle: str) -> UploadTicket:
        ticket = self._owned(org_id, upload_handle)
        if self.clock.now() >= ticket.expires_at:
            raise refuse(R.upload_expired, "the upload window has expired")
        if ticket.state is not UploadState.finalized:
            raise refuse(R.upload_not_finalized, f"upload is {ticket.state}")
        row = self.d.content.get(ticket.finalized.content_id)
        if row is None or row.state is not LifecycleState.live \
                or row.generation != ticket.finalized.generation:
            raise refuse(R.not_found, "the finalized source is gone")
        return ticket

    async def expire(self, limit: int) -> int:
        if isinstance(limit, bool) or not isinstance(limit, int) or limit < 1:
            raise errors.InvalidRequest("limit is a positive integer")
        async with self.d.lock:
            now = self.clock.now()
            due = sorted((t for t in self.d.tickets.values()
                          if t.state is UploadState.created and now >= t.expires_at),
                         key=lambda t: (t.expires_at, t.upload_handle))[:limit]
            for ticket in due:
                self._put(ticket, state=UploadState.expired)
            return len(due)

    # --- ReadinessStore ------------------------------------------------------------------
    async def admit_ready(self, request, idem, expectation: AdmissionExpectation):
        if not isinstance(expectation, AdmissionExpectation):
            raise errors.InvalidRequest("an AdmissionExpectation from the runtime is required")
        credit = expectation.accounting_regime is AccountingRegime.credit
        async with self.d.lock:
            found = await self.jobs.lookup(request.org_id, idem) if idem.key else None
            if found is not None:
                return self._replayed(found[0], credit)
            sources = self._manifest(request)
            if credit:
                pins, _card, _policy, _wallet = await self.jobs._credit_terms(request)
                if pins.rate_card_version != expectation.rate_card_version:
                    raise refuse(R.expectation_mismatch, "the pinned card is not this runtime's")
                serving = await self.jobs.catalog.serving_revision(pins.serving_version_id)
                if serving is None:
                    raise errors.NotFound("the pinned serving revision is not in the catalog")
                _check_pinned_capability(serving.capability, request)
                admission = await self.jobs.admit_credit(request, idem)
            else:
                serving = self._legacy_serving(request.model_revision)
                if serving is not None:             # a pre-catalog model: nothing to check
                    _check_pinned_capability(serving.capability, request)
                admission = await self.jobs.admit(request, idem, ())
            if admission.replayed:
                return self._replayed(admission, credit)
            now = self.clock.now()
            readiness = ExecutionReadiness(job_id=admission.request_id, org_id=admission.org_id,
                                           sources=sources, ready_at=admission.admitted_at)
            self.d.readiness[admission.request_id] = readiness
            for source in sources:
                self.d.references.setdefault(source.content_id, []).append(ContentReference(
                    content_id=source.content_id, generation=source.generation,
                    job_id=admission.request_id, org_id=admission.org_id, referenced_at=now))
            return admission, readiness

    def _legacy_serving(self, model_revision: str):
        """0019: a legacy job's `<alias>@<label>` names the serving revision of that label,
        if the catalog has one. ponytail: a scan of the v2 fake catalog's rows; a catalog
        without `servings` models no revision, so nothing is checked (a pre-catalog model)."""
        servings = getattr(getattr(self.jobs, "catalog", None), "servings", {})
        return next((serving for serving in servings.values()
                     if f"{serving.public_model_id}@{serving.revision_label}" == model_revision),
                    None)

    def _replayed(self, admission, credit: bool):
        if isinstance(admission, AdmissionV2) is not credit:
            raise errors.IdempotencyConflict("the key names a job of another regime")
        return admission, self.d.readiness.get(admission.request_id)

    def _manifest(self, request) -> tuple[ManifestSource, ...]:
        """Each ref -> a live source row of the request's organization, same digest."""
        if len({ref.handle for ref in request.media}) != len(request.media):
            raise refuse(R.invalid_manifest, "a manifest names each source once")
        sources = []
        for ref in request.media:
            row = self.d.content.get(self.d.by_key.get(
                (ContentLocation.object_store.value, ref.storage_ref), ""))
            if row is None or row.identity.org_id != request.org_id \
                    or row.identity.kind is not ContentKind.source \
                    or row.identity.digest != ref.digest or ref.org_id != request.org_id \
                    or row.state is LifecycleState.deleted:
                raise refuse(R.not_found, f"source {ref.handle} is not this org's content")
            if row.state is LifecycleState.tombstoned:
                raise refuse(R.content_retiring, f"source {ref.handle} is being deleted")
            if ref.kind is MediaKind.upload:
                ticket = self._resolve(request.org_id, ref.handle)
                if ticket.finalized.content_id != row.content_id:
                    raise refuse(R.not_found, "the upload names other content")
            sources.append(ManifestSource(content_id=row.content_id, generation=row.generation,
                                          ref=ref))
        return tuple(sources)

    async def readiness(self, job_id: str) -> ExecutionReadiness | None:
        return self.d.readiness.get(job_id) if ids.is_request_id(job_id) else None

    async def claim_preparation(self, job_id: str, worker_id: str):
        if job_id not in self.d.readiness:
            raise refuse(R.not_ready, f"job {job_id} has no execution-ready marker")
        return await self.jobs.claim_preparation(job_id, worker_id)

    # --- ContentLifecycle ----------------------------------------------------------------
    def _job_live(self, job_id: str, now: datetime,
                  kind: ContentKind = ContentKind.source) -> bool:
        """Non-terminal, or terminal and before its retain_until - set once, here, from the
        job's persisted facts (never recomputed): a result lasts exactly until the outcome's
        persisted `result_expires_at` (a job without one keeps no result), anything else
        the configured serving retention after settlement."""
        job = self.jobs.jobs.get(job_id)
        if job is None:
            return False
        if job.outcome is None:
            return True
        outcome = job.outcome
        until = self.d.retain.setdefault((job_id, kind), (
            outcome.result_expires_at or outcome.settled_at if kind is ContentKind.result
            else outcome.settled_at + timedelta(seconds=self.retention_s)))
        return now < until

    def _referenced(self, row: ContentObject, now: datetime) -> bool:
        identity = row.identity
        if any(ref.generation == row.generation and self._job_live(ref.job_id, now)
               for ref in self.d.references.get(row.content_id, ())):
            return True
        if identity.job_id is not None and self._job_live(identity.job_id, now, identity.kind):
            return True
        if identity.kind is ContentKind.upload_destination:
            ticket = self.d.tickets.get(identity.upload_handle)
            return ticket is not None and ticket.org_id == identity.org_id \
                and ticket.state is UploadState.created and now < ticket.expires_at
        return any(t.finalized is not None and t.finalized.content_id == row.content_id
                   and t.finalized.generation == row.generation and now < t.expires_at
                   for t in self.d.tickets.values())

    def _save(self, row: ContentObject, **changes) -> ContentObject:
        row = ContentObject.model_validate({**row.model_dump(), **changes})
        self.d.content[row.content_id] = row
        return row

    def _register(self, identity: ContentIdentity) -> ContentObject:
        key = (identity.location.value, identity.object_key)
        now = self.clock.now()
        fresh = dict(state=LifecycleState.live, registered_at=now,
                     eligible_at=now + timedelta(seconds=self.grace_s), claim=None,
                     tombstoned_at=None, deleted_at=None)
        row = self.d.content.get(self.d.by_key.get(key, ""))
        if row is None:
            row = ContentObject(content_id=self.ids.uuid(), generation=1, identity=identity,
                                **fresh)
            self.d.by_key[key] = row.content_id
            return self._save(row)
        if row.identity.org_id != identity.org_id:
            raise refuse(R.not_found, "the key is not this organization's")
        if row.state is LifecycleState.tombstoned:
            raise refuse(R.content_retiring, "a delete of this key is not yet acknowledged")
        if row.state is LifecycleState.deleted:
            return self._save(row, generation=row.generation + 1, identity=identity, **fresh)
        if None not in (row.identity.digest, identity.digest) \
                and row.identity.digest != identity.digest:
            raise refuse(R.bytes_changed, "the key already names other bytes")
        if identity.origin is ContentOrigin.written and row.state is LifecycleState.live \
                and row.identity.digest is not None and row.identity.digest == identity.digest:
            # M6 WR-7 (D10 0022): the runtime rewrites these bytes for a new request, so the
            # grace restarts (never shortens); a `discovered` registration never refreshes.
            return self._save(row, eligible_at=max(row.eligible_at,
                                                   now + timedelta(seconds=self.grace_s)))
        return row

    async def register(self, identity: ContentIdentity) -> ContentObject:
        async with self.d.lock:
            return self._register(identity)

    async def references(self, content_id: str) -> tuple[ContentReference, ...]:
        async with self.d.lock:
            if content_id not in self.d.content:
                raise refuse(R.not_found, "no such content")
            refs = self.d.references.get(content_id, ())
            for ref in refs:
                self._job_live(ref.job_id, self.clock.now())    # persists retain_until once
            return tuple(ContentReference.model_validate(
                {**ref.model_dump(), "retain_until": self.d.retain.get(
                    (ref.job_id, ContentKind.source))})
                for ref in refs)

    def _claimable(self, row: ContentObject, now: datetime) -> bool:
        claim_free = row.claim is None or now >= row.claim.expires_at
        if row.state is LifecycleState.tombstoned:
            return claim_free
        return row.state is LifecycleState.live and claim_free and now >= row.eligible_at \
            and not self._referenced(row, now)

    async def candidates(self, *, after: str | None, limit: int) -> ContentPage:
        if isinstance(limit, bool) or not isinstance(limit, int) or limit < 1:
            raise errors.InvalidRequest("limit is a positive integer")
        limit = min(limit, MAX_PAGE)
        start = _decode(after) if after is not None else None
        async with self.d.lock:
            now = self.clock.now()
            rows = sorted((r for r in self.d.content.values() if self._claimable(r, now)),
                          key=lambda r: (r.eligible_at, r.content_id))
            if start is not None:
                rows = [r for r in rows if (r.eligible_at, r.content_id) > start]
            page = tuple(rows[:limit])
            more = len(rows) > limit
            return ContentPage(items=page, next_cursor=_encode(page[-1]) if more else None)

    async def claim(self, content_id: str, generation: int, holder: str) -> DeletionClaim:
        async with self.d.lock:
            row = self.d.content.get(content_id)
            if row is None:
                raise refuse(R.not_found, "no such content")
            now = self.clock.now()
            if row.generation != generation or row.state is LifecycleState.deleted:
                raise refuse(R.claim_lost, "another generation of this object")
            if row.claim is not None and now < row.claim.expires_at:
                raise refuse(R.claim_held, "an unexpired claim holds this object")
            if row.state is LifecycleState.live:
                self._recheck(row, now)
            claim = DeletionClaim(content_id=content_id, generation=generation,
                                  fence=(row.claim.fence if row.claim else 0) + 1, holder=holder,
                                  claimed_at=now,
                                  expires_at=now + timedelta(seconds=self.claim_ttl_s))
            self._save(row, claim=claim)
            return claim

    def _recheck(self, row: ContentObject, now: datetime) -> None:
        if now < row.eligible_at:
            raise refuse(R.not_eligible, "inside the persisted grace")
        if self._referenced(row, now):
            raise refuse(R.reference_live, "a live reference protects this object")

    async def tombstone(self, claim: DeletionClaim) -> Tombstone:
        async with self.d.lock:
            row = self.d.content.get(claim.content_id)
            if row is None:
                raise refuse(R.not_found, "no such content")
            now = self.clock.now()
            if row.generation != claim.generation or row.claim is None \
                    or row.claim.fence != claim.fence or now >= row.claim.expires_at:
                raise refuse(R.claim_lost, "the claim is not the current, unexpired fence")
            if row.state is LifecycleState.live:
                self._recheck(row, now)
                row = self._save(row, state=LifecycleState.tombstoned, tombstoned_at=now)
            return Tombstone(content_id=row.content_id, generation=row.generation,
                             fence=claim.fence, location=row.identity.location,
                             object_key=row.identity.object_key, tombstoned_at=row.tombstoned_at)

    async def acknowledge_delete(self, tombstone: Tombstone) -> ContentObject:
        async with self.d.lock:
            row = self.d.content.get(tombstone.content_id)
            if row is None:
                raise refuse(R.not_found, "no such content")
            if row.generation != tombstone.generation or row.state is LifecycleState.deleted:
                return row          # an older generation's delete, or a repeat: no change
            if row.state is not LifecycleState.tombstoned or row.claim is None \
                    or row.claim.fence != tombstone.fence:
                raise refuse(R.claim_lost, "a newer claim finishes this delete")
            return self._save(row, state=LifecycleState.deleted, deleted_at=self.clock.now())


def _encode(row: ContentObject) -> str:
    raw = json.dumps([row.eligible_at.isoformat(), row.content_id]).encode()
    return "cc1." + base64.urlsafe_b64encode(raw).decode().rstrip("=")


def _decode(cursor: object) -> tuple[datetime, str]:
    try:
        if not isinstance(cursor, str) or not cursor.startswith("cc1."):
            raise ValueError(cursor)
        body = cursor[4:] + "=" * (-len(cursor[4:]) % 4)
        at, content_id = json.loads(base64.urlsafe_b64decode(body))
        return datetime.fromisoformat(at), ids.require_request_id(content_id)
    except (ValueError, TypeError):
        raise errors.InvalidCursor("not a content cursor this store issued") from None


def _check_pinned_capability(capability, request) -> None:
    """0019 `check_pinned_capability`, part for part: a string content is text, a
    `video_url` part video, any other part its own type, and media present needs video - all
    within the PINNED revision's `input_modalities` (`unsupported_media`, `param=messages`);
    a stream needs `stream_output` (`unsupported_parameter`, `param=stream`)."""
    needed = {"video"} if request.media else set()
    for message in request.messages:
        content = message.get("content")
        if isinstance(content, str):
            needed.add("text")
        elif isinstance(content, (list, tuple)):
            needed.update("video" if part.get("type") == "video_url" else part.get("type")
                          for part in content if isinstance(part, dict) and part.get("type"))
    if not needed <= set(capability.input_modalities):
        raise errors.UnsupportedMedia("the model does not accept this input modality",
                                      param="messages")
    if request.execution_mode is ExecutionMode.stream and not capability.stream_output:
        raise errors.UnsupportedParameter("the model does not stream its output",
                                          param="stream")
